### PR TITLE
XRT-1168: new sim features and enhanced event decoding/encoding

### PR DESCRIPTION
**Pull request by FraserG01**
- Added new abilities for simulator: Register Recipients, Generate Events, Create Notification Classes

- Added new decode and encode abilities for new event types:
EVENT_COMMAND_FAILURE
EVENT_ACCESS_EVENT
EVENT_DOUBLE_OUT_OF_RANGE
EVENT_SIGNED_OUT_OF_RANGE
EVENT_UNSIGNED_OUT_OF_RANGE
EVENT_CHANGE_OF_CHARACTERSTRING
EVENT_CHANGE_OF_TIMER

### CODE REVIEW

File: ports/esp32/src/device.c [hunk 1]
```diff
@@ -844,6 +844,16 @@ void Device_getCurrentDateTime(BACNET_DATE_TIME *DateTime)
...
 
+void Device_getCurrentTime(BACNET_TIME *Time)
+{
+    Update_Current_Time();
+
+    Time->hour = Local_Time.hour;
+    Time->min = Local_Time.min;
+    Time->sec = Local_Time.sec;
+    Time->hundredths = Local_Time.hundredths;
+}
```
Comment (dpazj) [inline ports/esp32/src/device.c:855]: Is this duplicated code required? (Duplicate from apps/piface/device.c) line 843

File: src/bacnet/basic/object/client/device-client.c [hunk 2]
```diff
@@ -698,6 +698,16 @@ void Device_getCurrentDateTime(BACNET_DATE_TIME *DateTime)
     DateTime->time = Local_Time;
 }
 
+void Device_getCurrentTime(BACNET_TIME *Time)
```
Comment (dpazj) [inline src/bacnet/basic/object/client/device-client.c:701]: Same comment about duplicate code

File: src/bacnet/basic/object/device.c [hunk 3]
```diff
@@ -990,6 +990,16 @@ void Device_getCurrentDateTime(BACNET_DATE_TIME *DateTime)
     DateTime->time = Local_Time;
 }
 
+void Device_getCurrentTime(BACNET_TIME *Time)
```
Comment (dpazj) [inline src/bacnet/basic/object/device.c:993]: Same comment about duplicate code

File: src/bacnet/basic/object/nc.c [hunk 4]
```diff
@@ -979,4 +972,72 @@ void Notification_Class_find_recipient(void)
...
 }
+
+void Notification_Class_Register_Destination (uint32_t device_id, uint16_t nc_instance, bool confirmed)
+{
+    bool empty_space = false;
+    bool already_registered = false;
+    BACNET_DESTINATION *RecipientEntry;
+    if (nc_instance >= NC_Info_Size)
+    {
+        fprintf(stdout, "NC instance is larger than maximum in the Simulator\n");
```
Comment (dpazj) [inline src/bacnet/basic/object/nc.c:983]: Do we not use the stack for more than just the simulator? Would this logging be an issue?

File: src/bacnet/event.c [hunk 5]
```diff
@@ -376,10 +450,190 @@ int event_notify_encode_service_request(
...
+                        apdu_len += len;
+                        break;
                     case EVENT_EXTENDED:
-                    case EVENT_COMMAND_FAILURE:
+                    case EVENT_CHANGE_OF_STATUS_FLAGS:
+                    case EVENT_CHANGE_OF_RELIABILITY:
+                    case EVENT_CHANGE_OF_DISCRETE_VALUE:
                     default:
-                        assert(0);
+                        //assert(0);
```
Comment (dpazj) [inline src/bacnet/event.c]: Is this meant to be commented out?

File: src/bacnet/event.c [hunk 6]
```diff
@@ -859,7 +1419,9 @@ int event_notify_decode_service_request(
                 break;
         }
     }
-
+    /**********************************************************************************/
+    /**********************************************************************************/
+    /**********************************************************************************/
```
Comment (dpazj) [inline src/bacnet/event.c]: Do we need this?

**Review by dpazj — CHANGES_REQUESTED** [review]

File: ports/esp32/src/device.c (hunk 1 above)
Comment (FraserG01) [reply ports/esp32/src/device.c:855]: Thought I would keep any stack additions consistent across platforms

File: src/bacnet/basic/object/client/device-client.c (hunk 2 above)
Comment (FraserG01) [reply src/bacnet/basic/object/client/device-client.c:701]: Thought I would keep any stack additions consistent across platforms

File: src/bacnet/event.c (hunk 6 above)
Comment (FraserG01) [reply src/bacnet/event.c]: nope, artefact from a newer version of the file found on the master branch of the stack (new event types supported)

File: src/bacnet/event.c (hunk 5 above)
Comment (FraserG01) [reply src/bacnet/event.c]: yeah, dont want to kill the program if a unsupported event is attempted to be generated, will just take the line out

File: src/bacnet/basic/object/device.c (hunk 3 above)
Comment (FraserG01) [reply src/bacnet/basic/object/device.c:993]: Thought I would keep any stack additions consistent across platforms

File: src/bacnet/basic/object/nc.c (hunk 4 above)
Comment (FraserG01) [reply src/bacnet/basic/object/nc.c:983]: We will not use this function on the client side of the stack. Most of this file will not be built in the client side of the stack anyway as its within:

#if defined(INTRINSIC_REPORTING)

which is not set

**Review by dpazj — APPROVED** [review]
LGTM

**Pull request merged**